File: 20_08_August/0811_12904_가장_긴_팰린드롬/Solution.java

class Solution
{

    private boolean check(String s, int start, int end) {
        int length = (end - start) / 2;

        for (int i = 0; i <= length; i++) {
            if (s.charAt(start + i) != s.charAt(end - 1 - i)) {
                return false;
            }
        }

        return true;
    }

    public int solution(String s)
    {
        int answer = 1;

        for (int i = 0; i < s.length(); i++) {

            for (int j = answer + 1; j <= s.length() - i; j++) {
                if (check(s, i, i + j)) {
                    answer = j;
                }
            }

        }

        return answer;
    }
}

File: 20_08_August/0811_12904_가장_긴_팰린드롬/Solution.py
def solution(s):
    answer = 0

    # tmp의 왼쪽, 오른쪽 대칭 검사
    def check(s, start, end):

        l = (end - start) // 2

        for i in range(l):

            if s[start + i] != s[end - 1 - i]:
                return False

        return True

    for i in range(len(s)):

        # 적어도 answer보다 큰 경우만 찾아야한다.
        for j in range(answer + 1, len(s) - i + 1):
            if check(s, i, i + j):
                answer = j
    
    return answer

print(solution("zxcvaaaa"))
print(solution('abcdcba'))

print(solution("abacde"))

print(solution("aaaa"))
